### ISSUE TITLE
Whether there is an example of C language is used for the firmware update

### ISSUE THREAD

**Issue by zmsoft** [Azure/azure-iot-sdk-c#296]
- **OS and version used:** <VERSION> 

- **SDK version used:** <VERSION> 

# Description of the issue:

# Code sample exhibiting the issue:

# Console log of the issue:

**Comment by ewertons** [comment]
Hi @zmsoft ,
Have you searched the list of issues?

**Comment by ewertons** [comment]
Example, https://github.com/Azure/azure-iot-sdk-c/issues/131

**Comment by zmsoft** [comment]
@ewertons thank you

**Comment by zmsoft** [comment]
@ewertons 
https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_mqtt_dm/pi_device/readme.md
Setup your IoT Hub and provision a device
(https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-device-management-get-started#create-a-device-management-enabled-iot-hub)
The page you requested could not be found.

**Comment by zmsoft** [comment]
@ewertons 
Info: Initialize Platform
Info: Instantiate the device.
Info: Initialize From Connection String.
Info: IoT Hub SDK for C, version 1.1.27
Info: Device successfully connected.
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR



When I run, the console prompts this error message.
ubuntu 14.04
IoT Hub SDK for C, version 1.1.27
iothub-explorer create mfirmwaretest --connection-string

**Comment by ewertons** [comment]
We will track the broken documentation on the separate item we created.

**Comment by ewertons** [comment]
@zmsoft ,
most systems will need to have the TLS certificates explicitly provided to the Azure IoT Client in order for the connection to truly succeed.

After the client gets [created](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_mqtt_dm/iothub_client_sample_mqtt_dm.c#L369), could you set the TLS certs similar* as it is done in [here](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_mqtt/iothub_client_sample_mqtt.c#L160) and see?

* use IoTHubClient_SetOption instead and call it directly (no need to add the surrounding #ifdef)

**Comment by zmsoft** [comment]
@ewertons 
~/Project/iot/fota/azure-iot-sdk-c/cmake/iothub_client/samples/iothub_client_sample_mqtt$ sudo ./iothub_client_sample_mqtt 
Info: IoT Hub SDK for C, version 1.1.27
IoTHubClient_LL_SetMessageCallback...successful.
IoTHubClient_LL_SendEventAsync accepted message [0] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [1] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [2] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [3] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [4] for transmission to IoT Hub.
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR
Info: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Info: Closing tlsio from a state other than TLSIO_STATE_EXT_OPEN or TLSIO_STATE_EXT_ERROR

When I test iothub_client_sample_mqtt
 I have the same error message
ubuntu 14.04
IoT Hub SDK for C, version 1.1.27
iothub-explorer create mfirmwaretest --connection-string



iothub_client_sample_mqtt_dm.c  

LogInfo("Initialize From Connection String.");
IOTHUB_CLIENT_HANDLE iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString, MQTT_Protocol);
                if (iotHubClientHandle == NULL)
                {
                    LogError("iotHubClientHandle is NULL!");
                    retValue = -7;
                }
                else
                {
                    LogInfo("Device successfully connected.");
                    if (IoTHubClient_SetOption(iotHubClientHandle, "logtrace", &traceOn) != IOTHUB_CLIENT_OK)
                    {
                        LogError("failed to set logtrace option");
                    }

                    #ifdef SET_TRUSTED_CERT_IN_SAMPLES
                    // For mbed add the certificate information
                    if (IoTHubClient_SetOption(iotHubClientHandle, OPTION_TRUSTED_CERT, certificates) != IOTHUB_CLIENT_OK)
                    {
                        printf("failure to set option \"TrustedCerts\"\r\n");
                    }
                    #endif // SET_TRUSTED_CERT_IN_SAMPLES


                    PHYSICAL_DEVICE *physical_device = physical_device_new(iot_device);
                    if (physical_device == NULL)
                    {
                        LogError("failed to make an iot device callback structure");
                        retValue = -8;
                    }


If you do not add the surrounding #ifdef compiler will be wrong

[ 84%] Built target serializer
Scanning dependencies of target iothub_client_sample_mqtt_dm
[ 85%] Building C object iothub_client/samples/iothub_client_sample_mqtt_dm/CMakeFiles/iothub_client_sample_mqtt_dm.dir/iothub_client_sample_mqtt_dm.c.o
/home/agreeya/Project/iot/fota/azure-iot-sdk-c/iothub_client/samples/iothub_client_sample_mqtt_dm/iothub_client_sample_mqtt_dm.c: In function ‘iothub_client_sample_mqtt_dm_run’:
/home/agreeya/Project/iot/fota/azure-iot-sdk-c/iothub_client/samples/iothub_client_sample_mqtt_dm/iothub_client_sample_mqtt_dm.c:393:68: error: ‘OPTION_TRUSTED_CERT’ undeclared (first use in this function)
                     if (IoTHubClient_SetOption(iotHubClientHandle, OPTION_TRUSTED_CERT, certificates) != IOTHUB_CLIENT_OK)
                                                                    ^
/home/agreeya/Project/iot/fota/azure-iot-sdk-c/iothub_client/samples/iothub_client_sample_mqtt_dm/iothub_client_sample_mqtt_dm.c:393:68: note: each undeclared identifier is reported only once for each function it appears in
/home/agreeya/Project/iot/fota/azure-iot-sdk-c/iothub_client/samples/iothub_client_sample_mqtt_dm/iothub_client_sample_mqtt_dm.c:393:89: error: ‘certificates’ undeclared (first use in this function)
                     if (IoTHubClient_SetOption(iotHubClientHandle, OPTION_TRUSTED_CERT, certificates) != IOTHUB_CLIENT_OK)
                                                                                         ^
make[2]: *** [iothub_client/samples/iothub_client_sample_mqtt_dm/CMakeFiles/iothub_client_sample_mqtt_dm.dir/iothub_client_sample_mqtt_dm.c.o] Error 1
make[1]: *** [iothub_client/samples/iothub_client_sample_mqtt_dm/CMakeFiles/iothub_client_sample_mqtt_dm.dir/all] Error 2
make: *** [all] Error 2



https://github.com/Azure/azure-iot-sdk-c/blob/ee3d90463909d8ce8b50f2e8ccc681722bd043c7/doc/Iothub_sdk_options.md
“Available Options”

"TrustedCerts" | OPTION_TRUSTED_CERT | const char* | Azure Server certificate used to validate TLS connection to iothub

Where can I get this server certificate and how to set it correctly?
many thanks

**Comment by ewertons** [comment]
@zmsoft , by adding the #ifdef it most likely disabled that code block.

Did you add the certs.h to your project? the certificates variable is defined there.
It's in the same sample suggested above.
https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_mqtt/iothub_client_sample_mqtt.c#L21

Do not add the checks for SET_TRUSTED_CERT_IN_SAMPLES, so the option can be set on your platform.

**Comment by zmsoft** [comment]
Yes, this header file was included.
-----------------------------------------------------------------------------------------------------------------
iothub_client_sample_mqtt/CMakeLists.txt

#Copyright (c) Microsoft. All rights reserved.
#Licensed under the MIT license. See LICENSE file in the project root for full license information.

#this is CMakeLists.txt for iothub_client_sample_mqtt

if(NOT ${use_mqtt})
    message(FATAL_ERROR "iothub_client_sample_mqtt being generated without mqtt support")
endif()

compileAsC99()

set(iothub_client_sample_mqtt_c_files
    iothub_client_sample_mqtt.c
)

set(iothub_client_sample_mqtt_h_files
    iothub_client_sample_mqtt.h
)

IF(WIN32)
    #windows needs this define
    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
    add_definitions(-DGB_MEASURE_MEMORY_FOR_THIS -DGB_DEBUG_ALLOC)
ENDIF(WIN32)

#Conditionally use the SDK trusted certs in the samples
#if(${set_trusted_cert_in_samples})
#    add_definitions(-DSET_TRUSTED_CERT_IN_SAMPLES)
#	include_directories(${PROJECT_SOURCE_DIR}/certs)
#	set(samples_cert_file ${PROJECT_SOURCE_DIR}/certs/certs.c)
#endif()

add_definitions(-DSET_TRUSTED_CERT_IN_SAMPLES)
include_directories(${PROJECT_SOURCE_DIR}/certs)
set(samples_cert_file ${PROJECT_SOURCE_DIR}/certs/certs.c)

include_directories(.)

add_executable(iothub_client_sample_mqtt ${iothub_client_sample_mqtt_c_files} ${iothub_client_sample_mqtt_h_files} ${samples_cert_file})

target_link_libraries(iothub_client_sample_mqtt  
    iothub_client_mqtt_transport
    iothub_client 
)
-------------------------------------------------------------------------------------------------------------------------
I opened the certificate related code.

Results:
Info: IoT Hub SDK for C, version 1.1.27
IoTHubClient_LL_SetMessageCallback...successful.
IoTHubClient_LL_SendEventAsync accepted message [0] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [1] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [2] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [3] for transmission to IoT Hub.
IoTHubClient_LL_SendEventAsync accepted message [4] for transmission to IoT Hub.
-> 15:37:12 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: iottest123.azure-devices.cn/mfirmwaretest/api-version=2016-11-14&DeviceClientType=iothubclient%2f1.1.27%20(Linux%3b%20x86_64) | PWD: XXXX | CLEAN: 0
<- 15:37:12 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
-> 15:37:12 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: devices/mfirmwaretest/messages/devicebound/# | QOS: 1
-> 15:37:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/mfirmwaretest/messages/events/temperatureAlert=false&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2Fjson&%24.ce=utf-8 | PACKET_ID: 3 | PAYLOAD_LEN: 83
-> 15:37:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/mfirmwaretest/messages/events/temperatureAlert=false&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2Fjson&%24.ce=utf-8 | PACKET_ID: 4 | PAYLOAD_LEN: 83
-> 15:37:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/mfirmwaretest/messages/events/temperatureAlert=false&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2Fjson&%24.ce=utf-8 | PACKET_ID: 5 | PAYLOAD_LEN: 83
-> 15:37:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/mfirmwaretest/messages/events/temperatureAlert=false&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2Fjson&%24.ce=utf-8 | PACKET_ID: 6 | PAYLOAD_LEN: 83
-> 15:37:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/mfirmwaretest/messages/events/temperatureAlert=false&%24.cid=CORE_ID&%24.mid=MSG_ID&%24.ct=application%2Fjson&%24.ce=utf-8 | PACKET_ID: 7 | PAYLOAD_LEN: 83
<- 15:37:12 SUBACK | PACKET_ID: 2 | RETURN_CODE: 1
<- 15:37:15 PUBACK | PACKET_ID: 3
Confirmation[0] received for message tracking id = 0 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- 15:37:15 PUBACK | PACKET_ID: 4
Confirmation[1] received for message tracking id = 1 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- 15:37:15 PUBACK | PACKET_ID: 5
Confirmation[2] received for message tracking id = 2 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- 15:37:15 PUBACK | PACKET_ID: 6
Confirmation[3] received for message tracking id = 3 with result = IOTHUB_CLIENT_CONFIRMATION_OK
<- 15:37:16 PUBACK | PACKET_ID: 7
Confirmation[4] received for message tracking id = 4 with result = IOTHUB_CLIENT_CONFIRMATION_OK
-> 15:41:02 PINGREQ
<- 15:41:02 PINGRESP


How to modify the cmake file of iothub_client_sample_mqtt_dm?
many thanks

**Comment by ewertons** [comment]
So looks like you commented the ```if(${set_trusted_cert_in_samples})``` and it worked for you, right?

Why don't you just invoke cmake passing that argument?

```
cmake -Dset_trusted_cert_in_samples=ON <path to Azure IoT C SDK>
```

Why do you need to modify the Cmake file? to always set the trusted certs? They are not needed on some platforms, that's why they are optional. 
You can just create the solution invoking cmake as above.

**Comment by zmsoft** [comment]
Thank you so much for your help.The compiler was successful and runs normal.

Here is the log info：
zmsoft@Vostro-3400:~/Project/iot/fota/azure-iot-sdk-c/cmake/iothub_client/samples/iothub_client_sample_mqtt_dm$ sudo ./iothub_client_sample_mqtt_dm 
[sudo] password for agreeya: 
isService
device_run_service
pid > 0 and exit
pid < 0 umask
close out the standard file descriptors
zmsoft@Vostro-3400:~/Project/iot/fota/azure-iot-sdk-c/cmake/iothub_client/samples/iothub_client_sample_mqtt_dm$ 


I added some print logs，and finally stopped, not going down.
Then without any error, the console doesn't have any logs.

**Comment by ewertons** [comment]
Thanks @zmsoft ,
we are glad it is working for you now.
We will close this issue, but feel free to either reopen it if you need to follow up, or create a new one for any new issue you may need help with.

Thanks for using the Azure IoT SDK.
